flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/KafkaConsumerIterator.java

```diff
@@ -0,0 +1,304 @@
+		PartitionMetadata metadata;
+		do {
+			metadata = findLeader(hosts, port, topic, partition);
+			try {
+				Thread.sleep(waitOnEmptyFetch);
+			} catch (InterruptedException e) {
+				throw new InterruptedException("Establishing connection to Kafka failed");
+			}
+		} while (metadata == null);
+
+		if (metadata.leader() == null) {
+			throw new RuntimeException("Can't find Leader for Topic and Partition. (at " + hosts.get(0)
+					+ ":" + port);
+		}
+
+		String leadBroker = metadata.leader().host();
+		clientName = "Client_" + topic + "_" + partition;
+
+		consumer = new SimpleConsumer(leadBroker, port, 100000, 64 * 1024, clientName);
+	}
+
+	/**
+	 * Initializes a connection from the earliest available offset.
+	 */
+	public void initializeFromBeginning() throws InterruptedException {
+		initialize();
+		readOffset = getLastOffset(consumer, topic, partition,
+				kafka.api.OffsetRequest.EarliestTime(), clientName);
+
+		resetFetchResponse(readOffset);
+	}
+
+	/**
+	 * Initializes a connection from the latest available offset.
+	 */
+	public void initializeFromCurrent() throws InterruptedException {
+		initialize();
+		readOffset = getLastOffset(consumer, topic, partition,
+				kafka.api.OffsetRequest.LatestTime(), clientName);
+
+		resetFetchResponse(readOffset);
+	}
+
+	/**
+	 * Initializes a connection from the specified offset.
+	 *
+	 * @param offset Desired Kafka offset
+	 */
+	public void initializeFromOffset(long offset) throws InterruptedException {
+		initialize();
+		readOffset = offset;
+		resetFetchResponse(readOffset);
+	}
+
+
+	// --------------------------------------------------------------------------------------------
+	//  Iterator methods
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Convenience method to emulate iterator behaviour.
+	 *
+	 * @return whether the iterator has a next element
+	 */
+	public boolean hasNext() {
+		return true;
+	}
+
+	/**
+	 * Returns the next message received from Kafka as a
+	 * byte array.
+	 *
+	 * @return next message as a byte array.
+	 */
+	public byte[] next() {
+		return nextWithOffset().getMessage();
+	}
+
+	/**
+	 * Returns the next message and its offset received from
+	 * Kafka encapsulated in a POJO.
+	 *
+	 * @return next message and its offset.
+	 */
+	public MessageWithOffset nextWithOffset() {
+
+		synchronized (fetchResponse) {
+			while (!iter.hasNext()) {
+				resetFetchResponse(readOffset);
+				try {
+					Thread.sleep(waitOnEmptyFetch);
+				} catch (InterruptedException e) {
+					e.printStackTrace();
+				}
+			}
+
+			MessageAndOffset messageAndOffset = iter.next();
+			long currentOffset = messageAndOffset.offset();
+
+			while (currentOffset < readOffset) {
+				messageAndOffset = iter.next();
+				currentOffset = messageAndOffset.offset();
+			}
+
+			readOffset = messageAndOffset.nextOffset();
+			ByteBuffer payload = messageAndOffset.message().payload();
```

Comment:
The assumption here is that the Kafka connector reuses its byte buffer, so all returned data has to be copied into a new array?